url4fdf79ae2c_beckchr_juel_tgz-pJ8-de_odysseus_el_samples_xml_sax_XMELFilterJ8/src/de/odysseus/el/misc/TypeConverterImpl.java

```diff
@@ -0,0 +1,367 @@
+/*
+ * Copyright 2006-2009 Odysseus Software GmbH
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package de.odysseus.el.misc;
+
+import java.beans.PropertyEditor;
+import java.beans.PropertyEditorManager;
+import java.math.BigDecimal;
+import java.math.BigInteger;
+
+import javax.el.ELException;
+
+/**
+ * Type Conversions as described in EL 2.1 specification (section 1.17).
+ */
+public class TypeConverterImpl implements TypeConverter {
+	private static final long serialVersionUID = 1L;
+
+	protected Boolean coerceToBoolean(Object value) {
+		if (value == null || "".equals(value)) {
+			return Boolean.FALSE;
+		}
+		if (value instanceof Boolean) {
+			return (Boolean)value;
+		}
+		if (value instanceof String) {
+			return Boolean.valueOf((String)value);
+		}
+		throw new ELException(LocalMessages.get("error.coerce.type", value, value.getClass(), Boolean.class));
+	}
+
+	protected Character coerceToCharacter(Object value) {
+		if (value == null || "".equals(value)) {
+			return Character.valueOf((char)0);
+		}
+		if (value instanceof Character) {
+			return (Character)value;
+		}
+		if (value instanceof Number) {
+			return Character.valueOf((char)((Number)value).shortValue());
+		}
+		if (value instanceof String) {
+			return Character.valueOf(((String)value).charAt(0));
+		}
+		throw new ELException(LocalMessages.get("error.coerce.type", value, value.getClass(), Character.class));
+	}
+
+	protected BigDecimal coerceToBigDecimal(Object value) {
+		if (value == null || "".equals(value)) {
+			return BigDecimal.valueOf(0l);
+		}
+		if (value instanceof BigDecimal) {
+			return (BigDecimal)value;
+		}
+		if (value instanceof BigInteger) {
+			return new BigDecimal((BigInteger)value);
+		}
+		if (value instanceof Number) {
+			return new BigDecimal(((Number)value).doubleValue());
+		}
+		if (value instanceof String) {
+			try {
+				return new BigDecimal((String)value);
+			} catch (NumberFormatException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)